ddht/v5_1/alexandria/abc.py

```diff
@@ -57,11 +58,44 @@ async def send_pong(
     ) -> None:
         ...
 
+    @abstractmethod
+    async def send_find_nodes(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        *,
+        distances: Collection[int],
+        request_id: Optional[bytes] = None,
+    ) -> bytes:
+        ...
+
+    @abstractmethod
+    async def send_found_nodes(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        *,
+        enrs: Sequence[ENRAPI],
+        request_id: bytes,
+    ) -> int:
+        ...
+
     #
     # High Level Request/Response
     #
     @abstractmethod
-    async def ping(self, node_id: NodeID, endpoint: Endpoint,) -> PongMessage:
+    async def ping(self, node_id: NodeID, endpoint: Endpoint) -> PongMessage:
+        ...
+
+    @abstractmethod
+    async def find_nodes(
```

Comment:
Is there a chance you know why discv5 decided to change how `FindNodes` works? In v4 you send the target node id and the remote sends you the closest nodes, even if they're in different buckets. In v5 you ask for a specific bucket, and you have to write extra code to ask for additional buckets if it detects that few nodes were returned. v4 also allows you to target specific buckets by using specially crafted target node ids, I think the only advantage to v5 is that it allows you to target multiple non-contiguous buckets at once?

Depending on how you feel about this I would kind of prefer if Alexandria used the v4 approach. It seems like it's easier to program against and isn't any less powerful.

Reply:
I think that by sending the target you *leak* information about what you're looking for.  It appears the v5 approach makes some attacks harder.  cc @fjl who may be able to provide more information.

Reply:
In discv4, it was actually not possible to target any specific bucket because the parameter was the public key preimage. The actual target is hashed on the server side in v4. We decided to change this in v5 and eventually landed on the current approach. The discussion for this happened here: https://github.com/ethereum/devp2p/issues/79 https://github.com/ethereum/devp2p/issues/141

Reply:
The rationale for using log distances in FINDNODE is that it's much harder for an attacker to short-circuit your lookup. The disadvantage is that routing is slightly slower.

Reply:
Thanks for the quick response, and for the links! This rationale makes sense to me, that's a nasty attack and avoiding it is pretty compelling.